### PR TITLE
Performance troubleshooting guide

### PR DESCRIPTION
**Pull request by gvwilson**
First complete draft - thanks @flamber and @paoliniluis for comments and improvements.

### CODE REVIEW

File: docs/troubleshooting-guide/performance.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+### The Metabase instance is getting so much traffic that loading the HTML page is slow.
+
+This is pretty rare, since our front end is not very large and browsers cache our JavaScript, but it's easy to rule out.
+
+**How to detect this:** Nothing appears on the page (not even the controls), or elements appear one by one.
+
+**How to fix this:** Look at page load times in the browser console [FIXME: link]. If it's taking a long time to load our HTML, CSS, or JavaScript, check to see whether a proxy, firewall, or other network component is slowing things down.
+
+### Caching is disabled.
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:33]: Should we note that ad-hoc queries (unsaved questions) are not cached - nor is exports?

File: docs/troubleshooting-guide/performance.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+By default caching is disabled so that we always re-run every question. However, if your data is only being updated every few seconds or minutes, you will improve performance by enabling caching.
+
+**How to detect this:** Open the Admin Panel, go to "Settings", and look in the "Caching" tab to see whether caching is enabled or not.
+
+**How to fix this:** [This guide][admin-caching] explains how to change the minimum query duration (we cache anything that takes longer than that to run) and the maximum cache size (which is the total memory the cache will ever use). You may need to experiment with these values over several days to find the best balance. For example, if you increase the total memory used by the cache, you may also want to increase the memory that the JVM is allowed to use (discussed below).
+
+### The answer you want isn't cached.
+
+If you have just restarted Metabase, it  has to refresh all its caches, so performance on some questions will be slower the first time you view them. This only happens once per question, but if you have thousands of users, each with their own custom questions, you may see a lot of slowdowns occurring all at once.
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:43]: Cache is stored in the appdb, so they still work after restart - as long as the cache duration is still valid.
Each question (and any filter combination) is it's own query, so different users might be viewing the same question, but with different filters, so each will have to load once before it's cached. This is much more obvious with Sandboxing, where each user has their own query for each question.

File: docs/troubleshooting-guide/performance.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+**How to detect this:** Look at Metabase's logs, or in the server's logs, to see when it was last restarted.
+
+**How to fix this:** Patience is a virtue.
+
+### The database is overloaded by other traffic.
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:49]: Could easily be caused by someone opening a dashboard with 20+ long-running queries, which will then mean everyone else trying to query the database will have to wait until there's available connections to the database. This is outlined in the scaling article.

File: docs/troubleshooting-guide/performance.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+### The question itself is slow.
+
+Joining half a dozen tables, each with a few million rows, simply takes a lot of time. On the other hand, while we do our best to create fully-formed SQL queries from graphical questions, SQL snippets, and questions that use other questions as starting points, it's a hard problem---particularly across as many databases as we support. We also don't take advantage of every quirk of every backend database. For example, Redshift stores values in columns rather than rows: some queries that work well for row-oriented databases are slow on columns and vice versa.
+
+**How to detect this:**
+
+1. Run the same SQL that Metabase is running but using a different tool. For example, if you have created a question using the Notebook Editor and you are using Postgres as your database, you can view the SQL, copy it, and run it from the command line using `psql`. (If you have written the question in SQL you can just copy and paste it.) If the query runs noticeably faster this way than it does when you run it through Metabase, the problem is almost certainly one of the ones described above.
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:65]: I've seen multiple times, where someone was debugging performance and was running the query directly on the database vs through Metabase, which might have several network layers between.
Generally I recommend using DBeaver, since it's also JDBC based, which matches most of our drivers too, and then running the tool on the same location as Metabase (or SSH passthru), so it's easier to detect potential network bottlenecks.

File: docs/troubleshooting-guide/performance.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Low performance when using Metabase can also be caused by incorrect typing of columns, e.g., by storing a numeric value as a string.  When this happens, the query converts values on the fly each time the query is run.
+
+**How to detect this:** Even a handwritten query will be slow if it has to do this conversion every time. You can also spot this by looking at the raw data types of the columns being used in the query.
+
+**How to fix this:** Amend the database schema to store numbers as numbers, timestamps as timestamps, and so on, rather than as strings or other data types.
+
+### The Java Virtual Machine is using too much memory.
+
+We discuss this problem and its solutions in [this troubleshooting guide](./running.md).
+
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:86]: JVM doesn't really fit into this guide, since Out Of Memory would cause other types of problems
```suggestion
```

File: docs/troubleshooting-guide/performance.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+### Metabase is running on an under-powered machine.
+
+**How to detect this:** Checking the performance logs for the server where Metabase is running will tell you whether it is hitting CPU or memory limits. However, it's much more likely that the database itself is hitting its limits, so please check it first.
+
+**How to fix this:** Upgrade to a more powerful server, or one with more memory.
+
+Note: If you are running in the cloud, the servers we select are enough for 98% of customers: we use a 2-node cluster for all our in-house analytics and a 3-node cluster for customers that autoscales whenever performance requires it. If you are using our cloud option and the Metabase server really is the problem, you should consider deploying your own instance.
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:93]: I'm unsure if this note is even needed. If someone is using Metabase Cloud and experiencing problems, then it's likely some different problems than general "slowness". And would require more debugging from our side, since it's a managed service, so people would never know if they're hitting CPU or memory limits.

File: docs/troubleshooting-guide/performance.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+**How to detect this:** The individual questions load quickly when viewed on their own, but the dashboard loads or updates slowly.
+
+**How to fix this:** See [this article][faster-dashboards] for tips on making dashboards more performant.
+
+### The UI appears to freeze when saving a question that has not yet been run
+
+**How to detect this:** If you save a question that has not been executed, MB runs the question while saving, which can make the UI look frozen.
+
+**How to fix this:** This is [a bug](https://github.com/metabase/metabase/issues/14957) and we are working to fix it. Until it's corrected, the workaround is to run the question before saving it.
```
Comment (flamber) [inline docs/troubleshooting-guide/performance.md:107]: There's a greater chance that people are hitting another problem than that specific bug. More likely lack of appdb connections, full Jetty queue, small network hiccups, etc.

**Review by flamber — APPROVED** [review]
This is great - small notes, hints to Scaling article, and a little removal.

File: docs/troubleshooting-guide/performance.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+7. The back end forwards those results to the front end.
+
+8. The front end creates the HTML to display them.
+
+## Specific Problems
+
+Each of the steps described is a potential performance bottleneck:
+
+### The Metabase instance is getting so much traffic that loading the HTML page is slow.
```
[the remark this answers is not in the document]
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:25]: incorporated - thank you

File: docs/troubleshooting-guide/performance.md (hunk 1 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:33]: incorporated - thank you.

File: docs/troubleshooting-guide/performance.md (hunk 3 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:49]: incorporated.

File: docs/troubleshooting-guide/performance.md (hunk 4 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:65]: thanks - added

File: docs/troubleshooting-guide/performance.md (hunk 6 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:89]: incorporated

File: docs/troubleshooting-guide/performance.md (hunk 7 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:103]: incorporated

File: docs/troubleshooting-guide/performance.md (hunk 2 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:43]: incorporated

File: docs/troubleshooting-guide/performance.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
+## Performance
+
+In order to troubleshoot performance problems, you first need to understand what happens when a question or dashboard is updated in Metabase:
```
Comment (paoliniluis) [inline docs/troubleshooting-guide/performance.md:3]: created or updated

File: docs/troubleshooting-guide/performance.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+**How to fix this:** [This guide][admin-caching] explains how to change the minimum query duration (we cache anything that takes longer than that to run) and the maximum cache size for each query result. You may need to experiment with these values over several days to find the best balance.
+
+### The answer you want isn't cached.
+
+Cached values are stored in the application database, so they will still be there if Metabase restarts, provided that the cache duration is still valid. Each question (and any filter combination) is its own query, so if different users are viewing the same question with different filters, each will have to load once before it's cached. This is particularly obvious with [data sandboxing][data-sandboxing], because each user essentially has their own query for each question.
+
+**How to detect this:** Look at Metabase's logs, or in the server's logs, to see when it was last restarted.
+
+**How to fix this:** Patience is a virtue.
```
Comment (paoliniluis) [inline docs/troubleshooting-guide/performance.md:51]: While I agree, we should say that the query needs to run more than 2 times to Metabase can do an average of query run times to detect that a question should be cached

File: docs/troubleshooting-guide/performance.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+2. We also recommend running the query with a JDBC-based tool like [DBeaver][dbeaver], which is also Java-based and uses most of the same database drivers as Metabase. Using this, and running the test query from the same machine that Metabase is running on, can help you determine if the problem is a network bottleneck.
+
+3. If the query runs slowly when sent by another tool, see if you can write a SQL query that calculates the same result as the question you have built in Metabase, but does so more quickly.
+
+**How to fix this:* If the problem is the SQL we generate:
+
+1. Check if you have the most recent version of Metabase: we fix problems as they're reported, and updating Metabase may make your problem go away.
+2. You can use your SQL in place of the code we generate, and make its result available to people who prefer the Notebook Editor as a starting point for their questions. [LINK]
```
Comment (paoliniluis) [inline docs/troubleshooting-guide/performance.md:78]: Where does this link goes?

File: docs/troubleshooting-guide/performance.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+**How to detect this:** Even a handwritten query will be slow if it has to do this conversion every time. You can also spot this by looking at the raw data types of the columns being used in the query.
+
+**How to fix this:** Amend the database schema to store numbers as numbers, timestamps as timestamps, and so on, rather than as strings or other data types.
+
+### Metabase is running on an under-powered machine.
+
+**How to detect this:** Checking the performance logs for the server where Metabase is running will tell you whether it is hitting CPU or memory limits. However, it's much more likely that the database itself is hitting its limits, so please check it first.
+
+**How to fix this:** Upgrade to a more powerful server or one with more memory. If you are running in the cloud, the servers we select are enough for 98% of customers (including our own internal analytics). If you are using our cloud option and the Metabase server really is the problem, you should consider deploying your own instance---please contact us for support.
```
Comment (paoliniluis) [inline docs/troubleshooting-guide/performance.md:93]: I would rephrase this to something like: "If you would like this to be taken care for you along other things like backup, memory and more, please consider using Metabase Cloud"

File: docs/troubleshooting-guide/performance.md (hunk 10 above)
Comment (flamber) [reply docs/troubleshooting-guide/performance.md:51]: A query is cached on first run - not on average - if it takes more time than the duration defined in the cache settings.
But the sections was rewritten, so the "How to detect this" and "How to fix this" needs an update.

File: docs/troubleshooting-guide/performance.md (hunk 9 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:3]: added

File: docs/troubleshooting-guide/performance.md (hunk 10 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:51]: rewritten

File: docs/troubleshooting-guide/performance.md (hunk 12 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:93]: done

File: docs/troubleshooting-guide/performance.md (hunk 11 above)
Comment (gvwilson) [reply docs/troubleshooting-guide/performance.md:78]: Link added.
